community: Add recursive sitemap support to GitbookLoader with concurrent processing

andrasfe opened the pull request:
## Description:
Enhanced `GitbookLoader` to support recursive sitemap structures and asynchronous processing. The loader now recursively processes sitemap index files, following links to child sitemaps, and extracts all URLs to content pages. Also added async processing.

## Issue:
Fixes #30629 - GitbookLoader fails to process nested sitemaps

## Dependencies:
None added

Commit:
Fixed gitbook bug of not being able to handle recursive site maps, added async processing for speeding document loading

eyurtsev commented on libs/community/langchain_community/document_loaders/gitbook.py:
```diff
@@ -11,6 +12,8 @@ class GitbookLoader(WebBaseLoader):
 
     1. load from either a single page, or
     2. load all (relative) paths in the navbar.
+
```
@andrasfe The solution should be leaning into composition rather than inheritance. We shouldn't be overriding a lot of the implementation.

andrasfe replied:
Ah, my bad. Completely overlooked the WebBaseLoader functionality. I did the necessary changes, albeit sitemap processing requires custom code and hence the base class methods have been used where applicable only. Also, noticed an issue with the base class (WebBaseLoader) aload() method calls asyncio.run(self.fetch_all(urls)) inside its implementation. So alazy_load() is the one implemented in GitBookLoader.

eyurtsev commented:
Could we speed up web loader instead of making changes in gitbook loader?

Commit:
changed implementation to use WebBaseLoader functionality to the extent possible

andrasfe commented:
> Could we speed up web loader instead of making changes in gitbook loader?

You right - any optimizations should be performed in WebBaseLoader. Nevertheless, this is a more involved task as it has lots of dependencies, so perhaps this could be another ticket.

andrasfe commented:
@eyurtsev , can you please review when you got a chance? Thank you!

eyurtsev commented:
@andrasfe I'll review this tomorrow more carefully. 

But if there are no changes here that optimize specifically for gitbookloader, then either changes will need to be made in the parent abstraction or the implementation needs to be refactored before we can merge the code.

eyurtsev commented:
For context, could you provide a quick explanation of what this does that was not do-able with the web based loader in terms of concurrent processing?

andrasfe commented:
> For context, could you provide a quick explanation of what this does that was not do-able with the web based loader in terms of concurrent processing?

The main issue I addressed, as outlined in the ticket I opened, was that the original implementation could not handle hierarchical sitemaps—clearly a bug. I introduced the optimization without realizing that WebBaseLoader already supports asynchronous loading. While there’s  room for further optimization in this class, as you pointed out, that’s not the primary focus of this ticket or fix. Thank you!!

eyurtsev commented on libs/community/pyproject.toml:
```diff
@@ -56,7 +56,7 @@ test = [
 ]
 codespell = ["codespell<3.0.0,>=2.2.0"]
-test_integration = ["pytest-vcr<2.0.0,>=1.0.2", "vcrpy<7,>=6"]
+test_integration = ["pytest-vcr<2.0.0,>=1.0.2", "vcrpy<7,>=6", "lxml"]
```
You can remove `lxml` from here 

Instead use in the pytest marker in the unit tests

```
@pytest.mark.requires("lxml")
```

andrasfe replied:
Done.

andrasfe replied:
Remove dependency for unit tests and added it for integration.

eyurtsev commented on libs/community/langchain_community/document_loaders/gitbook.py:
```diff
@@ -62,23 +64,153 @@ def __init__(
+                try:
+                    # We need to temporarily set the web_paths to the sitemap URL
+                    original_web_paths = self.web_paths
+                    self.web_paths = [sitemap_url]
```
The mutation in place makes this code not thread safe?

andrasfe replied:
addressed this by creating a private method that creates a new WebLoader, `_create_web_loader()` per thread - hence reusing the same instance iteratively. I feel this is a cleaner approach than using thread synchronization.

eyurtsev commented on libs/community/tests/unit_tests/document_loaders/test_gitbook.py:
```diff
@@ -90,115 +67,309 @@ def test_lazy_load_with_custom_sitemap(
-    mock_scrape.return_value = mock_sitemap_soup
+    sitemap_changelog_pages_content = """
+    <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
+        <url><loc>https://example.com/changelog/update1</loc></url>
```
What happens if the URL points to https://localhost:8000/foo?

andrasfe replied:
Added unit test to validate change that addresses the CVE concerns: added an optional parameter, `allowed_domains` that whitelists accepted domain. This required in turn a modification to work with absolute url-s as opposed to relative ones. Added a unit test to validate the functionality.

eyurtsev commented on libs/community/langchain_community/document_loaders/gitbook.py:
```diff
@@ -10,7 +12,7 @@ class GitbookLoader(WebBaseLoader):
 
     1. load from either a single page, or
-    2. load all (relative) paths in the navbar.
+    2. load all (relative) paths in the sitemap, handling nested sitemap indexes.
```
The implementation seems too complex to me.

Why does GitbookLoader not inherit from DocumentLoader?

It can implement lazy_load and alazy_load and use WebBaseLoader as it needs to fetch content?

eyurtsev reviewed (commented):
Is it possible to make 3 changes to this:

1) Stop inherit from WebBaseLoader and re-use WebBaseLoader to simplify implementation?
2) Make sure that this does not introduce a CVE unless there's something already built in that can filter urls that are not in allowed domains.
3) remove changes to pyproject.toml and uv.lock and replace with pytest.marker.requires

Commit:
GitbookLoader now inherits from BaseLoader and uses WebLoader thread-safely, CVE support added, reverted pyproject.toml and uv.lock to master

andrasfe commented on libs/community/pyproject.toml:
```diff
@@ -7,8 +7,8 @@ authors = []
-    "langchain-core<1.0.0,>=0.3.45",
-    "langchain<1.0.0,>=0.3.21",
+    "langchain-core<1.0.0,>=0.3.51",
+    "langchain<1.0.0,>=0.3.23",
```
Done. while I made changes to pyproject.toml and uv.sync earlier, I reverted them to match upstream in my latest commit.

andrasfe replied:
1. Implementation now extends BalseLoader and implements lazy_lad and alazy_load in a thread-safe manner. Code is simpler, indeed.
2. As mentioned above, added optional init argument/member allowed_domains that ignores with warning any non-whitelisted domain. This required change to work strictly with absolute paths, given the possibility for existence of sitemaps pointing to external domains. If this argument is None, default behavior is to accept only documents from the domain of the constructor-passed web_page.
3. mad those changes to revert the 2 files and remove dependency on lxml. I updated these files from the latest in master origin, hence the changes are not mine.

andrasfe commented:
All changes implemented, as requested. thank you!

andrasfe commented:
Hi @eyurtsev , when you got a chance.... thank you :)

eyurtsev commented on libs/community/langchain_community/document_loaders/gitbook.py:
```diff
@@ -42,43 +46,304 @@ def __init__(
+            web_loader.web_paths = new_urls
+
+            # Use the same WebBaseLoader's ascrape_all for efficient parallel fetching
+            soups = await web_loader.ascrape_all(new_urls, parser="xml")
```
This is likely using the built-in `xml` library in python which is not a safe default.

https://docs.python.org/3/library/xml.html

can we specify `lxml` by default?
`lxml`

andrasfe replied:
fixed.

eyurtsev commented on libs/community/tests/unit_tests/document_loaders/test_gitbook.py:
```diff
@@ -1,11 +1,15 @@
 
 from langchain_community.document_loaders.gitbook import GitbookLoader
 
+# These tests mock the XML parsing functionality to avoid requiring lxml.
```
`pytest.mark.requires(["lxml"])` allows testing with LXML. 

from community root folder `git grep requires` for usage examples

andrasfe replied:
got it and fixed it.

eyurtsev reviewed (commented):
@andrasfe thank you for the updated implementation and apologies for the delay I just came back from vacation yesterday.

The implementation looks good overall, i flagged a few places that still need to be adjusted for security purposes.

We're in the process of moving community to a separate repository:

https://github.com/langchain-ai/langchain-community

Would you be able to address the issue and re-open the PR there?

Commit:
fix(gitbook): Enhance SSRF protection in GitbookLoader

- Improve domain validation to ensure URLs come from allowed domains
- Add explicit scheme validation to prevent protocol-based attacks
- Implement safe URL checking before adding URLs to processing queue
- Refactor URL handling code to apply validation consistently

andrasfe commented on libs/community/langchain_community/document_loaders/gitbook.py:
```diff
@@ -42,43 +46,304 @@ def __init__(
+        if self.allowed_domains is None:
+            return True
+
+        netloc = urlparse(url).netloc
```
existing implementation uses  `netloc = urlparse(url).netloc`, which extracts the full network location (like www.something.x or www.something.xy). If allowed_domains = {"www.something.x"}, the check "www.something.xy" in {"www.something.x"} will correctly return False. Added fix to validate the scheme, which was indeed not there. Also updated unit tests to validate functionality.

andrasfe commented on libs/community/langchain_community/document_loaders/gitbook.py:
```diff
@@ -42,43 +46,304 @@ def __init__(
+            soups = await web_loader.ascrape_all(new_urls, parser="xml")
+
+            # Restore original web_paths
+            web_loader.web_paths = original_web_paths
```
the suggestion to avoid in-place mutation by creating a web loader would indeed make the code cleaner and less likely for future changes to impact its functionality, it would, however, introduce a slight performance overhead, especially for some sites that have hundreds or even thousand nested pages. I left it this way for now but if you feel we should change it, will be happy to do so.

andrasfe commented:
Thank you @eyurtsev and no problem at all. I saw you had no activity during this time frame - know you were off. I completed the changes ad provided my 2 cents for some, but I'm open to further suggestions. I will open a PR against the new repo once I get confirmation from you that we are good to go.

ccurme commented:
Closing for now as we've already moved the library.

andrasfe commented:
re-created PR in the new repo: https://github.com/langchain-ai/langchain-community/pull/13
